### PR TITLE
Add api middleware

### PR DESCRIPTION
**Pull request by mvn-caonguyen-dn**
## Description
Add api middleware
## Reference Requirements
- Product Backlog ID: None
- Task ID: None

## Reference Isuees
#{issue_id}

## Migrations
NO

## Impacted Areas in Application
List general components of the application that this PR will affect:

- app/Http/Kernel.php
- resources/lang/en/api.php

## SQL query and `explain`
List of all queries and attach the result of each explaining query (in text, screenshot...)

### CODE REVIEW

**Review by ghost — COMMENTED** [review]
LGTM

File: routes/api.php [hunk 1]
```diff
@@ -23,4 +23,8 @@
     Route::get('users/{user}', 'UserController@show');
     Route::get('books/{id}/reviews', 'PostController@reviews');
     Route::get('comments', 'CommentController@comments');
+
+    Route::group(['middleware' => 'apiLogin'], function(){
```
Comment (at-huydq) [inline routes/api.php]: những route nào cần auth thì bỏ vào đây, example `Route::get('users/{user}', 'UserController@show');`

File: routes/api.php [hunk 2]
```diff
@@ -23,4 +23,8 @@
...
+        // Route::post('posts', 'PostController@store');
```
Comment (at-huydq) [inline routes/api.php]: ko dùng thì xoá đi nhé

File: app/Http/Middleware/Api/APILoginMiddleware.php [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+        $user = $accessToken ? User::where('access_token', $accessToken)->firstOrFail() : null;
+
+        if ($user) {
+            if (Carbon::parse($user->expires_at)->gt(Carbon::now())) {
+                Auth::login($user);
+                return $next($request);
+            }
+
+            throw new TokenMismatchException(__('api.error.session_expired'), Response::HTTP_NOT_FOUND);
```
Comment (at-huydq) [inline app/Http/Middleware/Api/APILoginMiddleware.php]: return về mã code 401

File: app/Http/Middleware/Api/APILoginMiddleware.php [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        }
+
+        throw new TokenMismatchException(__('api.error.token_not_found'), Response::HTTP_NOT_FOUND);
```
Comment (at-huydq) [inline app/Http/Middleware/Api/APILoginMiddleware.php]: return về message `Unauthorized` , code : 401

File: app/Http/Middleware/Api/APILoginMiddleware.php [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+     * @param \Illuminate\Http\Request $request request
+     * @param \Closure                 $next    next
+     *
+     * @return mixed
+     */
+    public function handle($request, Closure $next)
+    {
+        $accessToken = $request->headers->get('access-token');
+
+        $user = $accessToken ? User::where('access_token', $accessToken)->firstOrFail() : null;
```
Comment (at-huydq) [inline app/Http/Middleware/Api/APILoginMiddleware.php]: check first thôi, ko check firstOrFail;

File: app/Http/Middleware/Api/APILoginMiddleware.php [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+namespace App\Http\Middleware\Api;
+
+use App\Model\User;
+use Carbon\Carbon;
+use Illuminate\Support\Facades\Auth;
+use Illuminate\Http\Response;
+use Illuminate\Session\TokenMismatchException;
+use Closure;
+
+class APILoginMiddleware
```
Comment (at-huydq) [inline app/Http/Middleware/Api/APILoginMiddleware.php]: rename to `TokenAuthenticationMiddleware`

File: app/Http/Kernel.php [hunk 7]
```diff
@@ -58,5 +58,6 @@ class Kernel extends HttpKernel
         'guest' => \App\Http\Middleware\RedirectIfAuthenticated::class,
         'throttle' => \Illuminate\Routing\Middleware\ThrottleRequests::class,
         'admin' => \App\Http\Middleware\AdminLoginMiddleware::class,
+        'apiLogin' => \App\Http\Middleware\Api\APILoginMiddleware::class,
```
Comment (at-huydq) [inline app/Http/Kernel.php]: rename to `TokenAuthentication`

**Review by ghost — COMMENTED** [review]
please check my comment

**Comment by at-huydq** [comment]
LGTM.

**Pull request merged**